compiler/src/main/java/org/qbicc/type/definition/DefinedTypeDefinitionImpl.java

```diff
@@ -254,6 +425,88 @@ public LoadedTypeDefinition load() throws VerifyFailedException {
+        }
+        return parameters;
+    }
+
+    private void addMethodsToVisitedSet(final LoadedTypeDefinition ltd, final HashMap<String, HashSet<MethodDescriptor>> visited) {
+        if (ltd == null) {
+            return;
+        }
+        int cnt = ltd.getMethodCount();
+        for (int i = 0; i < cnt; i ++) {
+            addMethodToVisitedSet(ltd.getMethod(i), visited);
+        }
+        addMethodsToVisitedSet(ltd.getSuperClass(), visited);
+    }
+
+    /**
+     * Add one method to the visited set.
+     * @param method the method to add
+     * @param visited the visited set
+     * @return {@code true} if the method is not static or private and was not previously in the set, or {@code false} if no new method was added
+     */
+    private boolean addMethodToVisitedSet(final MethodElement method, final HashMap<String, HashSet<MethodDescriptor>> visited) {
+        if (method.isStatic() || method.isPrivate()) {
+            return false;
+        } else {
+            return visited.computeIfAbsent(method.getName(), DefinedTypeDefinitionImpl::newSet).add(method.getDescriptor());
+        }
+    }
+
+    private int findInterfaceImplementations(LoadedTypeDefinition[] interfaces, HashMap<String, HashSet<MethodDescriptor>> visited, HashMap<String, HashMap<MethodDescriptor, List<MethodElement>>> toAdd, int depth) {
+        int processed = 0;
+        for (int i = 0; i < interfaces.length; i ++) {
+            processed += findInterfaceImplementations(interfaces[i], visited, toAdd, depth);
+        }
+        return processed;
+    }
+
+    private int findInterfaceImplementations(LoadedTypeDefinition currentInterface, HashMap<String, HashSet<MethodDescriptor>> visited, HashMap<String, HashMap<MethodDescriptor, List<MethodElement>>> toAdd, int depth) {
+        if (depth > 0) {
+            return findInterfaceImplementations(currentInterface.getInterfaces(), visited, toAdd, depth - 1);
+        } else {
+            int cnt = currentInterface.getMethodCount();
+            for (int i = 0; i < cnt; i ++) {
+                MethodElement method = currentInterface.getMethod(i);
+                if (addMethodToVisitedSet(method, visited)) {
+                    // we didn't have this one before
+                    toAdd.computeIfAbsent(method.getName(), DefinedTypeDefinitionImpl::newMap).computeIfAbsent(method.getDescriptor(), DefinedTypeDefinitionImpl::newList).add(method);
+                } else {
+                    // check to see whether we need to add this one to our level
+                    HashMap<MethodDescriptor, List<MethodElement>> map1 = toAdd.get(method.getName());
+                    if (map1 != null) {
+                        List<MethodElement> list = map1.get(method.getDescriptor());
+                        if (list != null) {
+                            // we have another impl option for this
+                            list.add(method);
+                        }
+                    }
+                }
```

Comment:
For default methods, the first rule regarding using the default implementation is that "the class always wins" so we can skip creating the interface sets if there is a concrete implementation from the current class or its superclasses.

Reply:
The `toAdd` set should only include methods that were not already present on the class or the superclas (this is done on line 228).  Any method found to already be in the visited set is skipped.  This code right here adds all the possible default method implementations that we've found - but only if we didn't already have an implementation for that class.